cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/BasicWorkflowToken.java

```diff
@@ -27,19 +32,138 @@
  */
 public class BasicWorkflowToken implements WorkflowToken {
   private Map<String, Map<String, Long>> mapReduceCounters;
+  private final Map<String, List<NodeValueEntry>> tokenValueMap = Maps.newHashMap();
+  private String nodeName;
 
+  void setCurrentNode(String nodeName) {
+    this.nodeName = nodeName;
+  }
+
+  /**
+   * Merge the other WorkflowToken passed to the method as a parameter
+   * with the WorkflowToken on which the method is invoked.
+   * @param other the other WorkflowToken to be merged
+   */
+  void mergeToken(BasicWorkflowToken other) {
+    Map<String, List<NodeValueEntry>> otherTokenValueMap = other.tokenValueMap;
+
+    for (Map.Entry<String, List<NodeValueEntry>> entry : otherTokenValueMap.entrySet()) {
+      if (!tokenValueMap.containsKey(entry.getKey())) {
+        // Key is newly added to the other WorkflowToken
+        tokenValueMap.put(entry.getKey(), Lists.<NodeValueEntry>newArrayList());
+      }
+
+      // Iterate over the list of NodeValueEntry corresponding to the current key.
+      // Only add those NodeValueEntry to the merged token which already do not exist.
+      for (NodeValueEntry otherNodeValueEntry : otherTokenValueMap.get(entry.getKey())) {
+        boolean otherNodeValueEntryExist = false;
+        for (NodeValueEntry thisNodeValueEntry : tokenValueMap.get(entry.getKey())) {
+          if (thisNodeValueEntry.equals(otherNodeValueEntry)) {
+            otherNodeValueEntryExist = true;
+            break;
+          }
+        }
+        if (!otherNodeValueEntryExist) {
+          tokenValueMap.get(entry.getKey()).add(otherNodeValueEntry);
+        }
+      }
+    }
+    if (other.getMapReduceCounters() != null) {
+      setMapReduceCounters(other.getMapReduceCounters());
+    }
+  }
+
+  @Override
+  public void put(String key, String value) {
+    if (nodeName == null) {
+      throw new IllegalStateException("Node name cannot be null.");
+    }
+    List<NodeValueEntry> nodeValueList = tokenValueMap.get(key);
+    if (nodeValueList == null) {
+      nodeValueList = Lists.newArrayList();
+      tokenValueMap.put(key, nodeValueList);
+    }
+    nodeValueList.add(new NodeValueEntry(nodeName, value));
```

Comment:
oh ok. I think while inserting into the List, if the entry for the nodeName already exists for the specific key, that entry should be replaced by the new one. So that for any key, we will always have only one entry for the specific node. I will update the code accordingly.